### PR TITLE
EDX-40: MIN - Close a Message

### PR DESCRIPTION
**Pull request by TimOsse**

### CODE REVIEW

File: backend/src/components/edx/exchange.js [hunk 1]
```diff
@@ -238,6 +238,29 @@ async function markAs(req, res) {
...
+    return res.status(HttpStatus.UNAUTHORIZED).json({
+      message: 'No access token'
+    });
+  }
+  try {
+    const currentExchange = await getData(token, config.get('server:edx:exchangeURL') + `/${req.params.secureExchangeID}`);
+    currentExchange.secureExchangeStatusCode = 'CLOSED';
+    currentExchange.createDate = null;
+    currentExchange.updateDate = null;
+
```
Comment (arcshiftsolutions) [inline backend/src/components/edx/exchange.js:253]: Please add an additional check here that the logged in user has the correct permission & school access. You can use the `activeInstituteIdentitfier` & `activeInstitutePermissions` to confirm. Just check it against the `currentExchange` you pulled.

Comment (arcshiftsolutions) [inline backend/src/components/edx/exchange.js:253]: Please add a check here to confirm the user has the `SECURE_EXCHANGE` role and correct team access to make the change. Check the exchange for the ownership team and confirm the user has the access.

Comment (TimOsse) [reply backend/src/components/edx/exchange.js:274]: Please have another look. I have implemented permission validation requested.

File: backend/src/components/edx/exchange.js [hunk 2]
```diff
@@ -252,15 +252,19 @@ async function markAsClosed(req, res) {
...
     .then(async ([statusCodeResponse, ministryTeamCodeResponse, dataResponse]) => {
       if (statusCodeResponse && ministryTeamCodeResponse && dataResponse) {
+        const userInfo = utils.getUser(req);
+        let userRoles = userInfo.realm_access.roles;
         let roleCheckPass = false;
         let ownerTeamCheckPass = false;
         let exchange = dataResponse;
         if(exchange['ministryOwnershipTeamID']){
           let matchedMinTeam = ministryTeamCodeResponse.find(minstryTeam => minstryTeam['ministryOwnershipTeamId'] === exchange['ministryOwnershipTeamID']);
           if(matchedMinTeam['groupRoleIdentifier']) {
```
Comment (arcshiftsolutions) [inline backend/src/components/edx/exchange.js:262]: There's a missing condition here; we need to check if the user has the correct team role. Check if the `userRoles` object contains the `groupRoleIdentifier` you found. You can use a lambda function for this if you want.

File: backend/src/components/edx/exchange.js [hunk 3]
```diff
@@ -252,15 +252,19 @@ async function markAsClosed(req, res) {
...
             ownerTeamCheckPass = true;
-            if(matchedMinTeam['groupRoleIdentifier'] == 'PEN_TEAM_ROLE'){
-              roleCheckPass = true;
+            for (const role of userRoles) {
```
Comment (arcshiftsolutions) [inline backend/src/components/edx/exchange.js:264]: Suggest you use a lambda function for this loop as well.

File: backend/src/components/edx/exchange.js (hunk 2 above)
Comment (TimOsse) [reply backend/src/components/edx/exchange.js:262]: `userRoles` is just an array of roles. The userInfo object does not contain any team related data. I could call the getEdxUsers api, filter out the current user and use edxUserSchools[]->edxUserSchoolRoles[]->edxRoleCode. The other option I see is to use the groupRoleIdentifier field from the ministryTeamCodeResponse. What do you think?

**Review by arcshiftsolutions — APPROVED** [review]

**Pull request merged**